### PR TITLE
Issue 134,137: Fix BookKeeper errors

### LINKED ISSUE #134: Observing Bookie restart / CrashLoopBackOff with BookieException$InvalidCookieException
While running moderate IO workload (10 writters, 10 readers and 100 streams with 1000 bytes size and 100 events per sec) observed bookie restart or CrashLoopBackOff on PKS environment.

Environment details: PKS / K8 with medium cluster:
3 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
5 worker nodes @ xlarge.cpu(8 cpu, 8 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend

Pravega version: 0.5.0-2151.2e8f0b7
Zookeeper Operator : 0.2.1
Pravega Operator: 0.3.0-1

```
NAMESPACE     NAME                                                 READY     STATUS             RESTARTS   AGE
default       example-0                                            1/1       Running            0          1h
default       example-1                                            1/1       Running            0          1h
default       example-2                                            1/1       Running            0          1h
default       pravega-benchmark1                                   1/1       Running            0          50m
default       pravega-benchmark2                                   1/1       Running            0          44m
default       pravega-benchmark3                                   1/1       Running            0          44m
default       pravega-benchmark4                                   1/1       Running            0          44m
default       pravega-benchmark5                                   1/1       Running            0          43m
default       pravega-bookie-0                                     1/1       Running            0          1h
default       pravega-bookie-1                                     0/1       Running            8          1h
default       pravega-bookie-2                                     0/1       CrashLoopBackOff   10         1h
default       pravega-bookie-3                                     1/1       Running            0          1h
default       pravega-operator-577d9b49-qtjxn                      1/1       Running            0          1h
default       pravega-pravega-controller-58f7f5658b-56lww          1/1       Running            0          1h
default       pravega-pravega-controller-58f7f5658b-vcb8g          1/1       Running            0          1h
default       pravega-pravega-segmentstore-0                       1/1       Running            0          1h
default       pravega-pravega-segmentstore-1                       1/1       Running            0          59m
default       pravega-pravega-segmentstore-2                       1/1       Running            0          58m
```

Bookie Logs showing exception:
```
19-03-12 13:42:27,352 - INFO  - [main-SendThread(example-client.default.svc.cluster.local:2181):ClientCnxn$SendThread@1303] - Session establishment complete on server example-client.default.svc.cluster.local/10.100.200.114:2181, sessionid = 0x1001f770d67028d, negotiated timeout = 10000
2019-03-12 13:42:27,502 - INFO  - [main-EventThread:ZooKeeperWatcherBase@130] - ZooKeeper client is connected now.
2019-03-12 13:42:28,624 - ERROR - [main:Bookie@471] - There are directories without a cookie, and this is neither a new environment, nor is storage expansion enabled. Empty directories are [/bk/index/current]
2019-03-12 13:42:28,624 - INFO  - [main:BookieNettyServer@396] - Shutting down BookieNettyServer
2019-03-12 13:42:28,639 - ERROR - [main:Main@221] - Failed to build bookie server
org.apache.bookkeeper.bookie.BookieException$InvalidCookieException:
        at org.apache.bookkeeper.bookie.Bookie.checkEnvironmentWithStorageExpansion(Bookie.java:475)
        at org.apache.bookkeeper.bookie.Bookie.checkEnvironment(Bookie.java:257)
        at org.apache.bookkeeper.bookie.Bookie.<init>(Bookie.java:641)
        at org.apache.bookkeeper.proto.BookieServer.newBookie(BookieServer.java:131)
        at org.apache.bookkeeper.proto.BookieServer.<init>(BookieServer.java:100)
        at org.apache.bookkeeper.server.service.BookieService.<init>(BookieService.java:43)
        at org.apache.bookkeeper.server.Main.buildBookieServer(Main.java:299)
        at org.apache.bookkeeper.server.Main.doMain(Main.java:219)
        at org.apache.bookkeeper.server.Main.main(Main.java:201)```

### LINKED ISSUE #137: Observed Bookie restart with exception java.lang.OutOfMemoryError: Java heap space
While running moderate IO workload (10 writters, 10 readers and 100 streams with 1000 bytes size and 100 events per sec) observed one bookie (bokie3) restart with exception java.lang.OutOfMemoryError: Java heap space on PKS environment.

Environment details: PKS / K8 with medium cluster:
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend

Pravega version: 0.5.0-2151.2e8f0b7
Zookeeper Operator : 0.2.1
Pravega Operator: adrianmo/pravega-operator:issue-134

```
NAMESPACE     NAME                                                 READY     STATUS      RESTARTS   AGE
default       example-0                                            1/1       Running     0          2h
default       example-1                                            1/1       Running     0          2h
default       example-2                                            1/1       Running     0          2h
default       pravega-benchmark1                                   1/1       Running     0          1h
default       pravega-benchmark2                                   1/1       Running     0          1h
default       pravega-benchmark3                                   1/1       Running     0          1h
default       pravega-benchmark4                                   1/1       Running     0          1h
default       pravega-benchmark5                                   1/1       Running     0          1h
default       pravega-bookie-0                                     1/1       Running     0          1h
default       pravega-bookie-1                                     1/1       Running     0          1h
default       pravega-bookie-2                                     1/1       Running     0          1h
default       pravega-bookie-3                                     1/1       Running     1          1h
default       pravega-operator-f7d987f74-2mrft                     1/1       Running     0          1h
default       pravega-pravega-controller-58f7f5658b-mrvdz          1/1       Running     0          1h
default       pravega-pravega-controller-58f7f5658b-rpqfz          1/1       Running     0          1h
default       pravega-pravega-segmentstore-0                       1/1       Running     0          1h
default       pravega-pravega-segmentstore-1                       1/1       Running     0          1h
default       pravega-pravega-segmentstore-2                       1/1       Running     0          1h
default       shaka-zulu-nfs-client-provisioner-59d7f8f84c-wdh95   1/1       Running     0          2h
default       zookeeper-operator-5fcf9497f5-z5bnw                  1/1       Running     0          2h
```

Log:
```
2019-03-13 10:33:05,184 - INFO  - [ZKC-connect-executor-0:ZooKeeperClient$1$1@130] - ZooKeeper session 1000052f01f050a is created to 10.100.200.137:2181.
2019-03-13 10:33:16,839 - WARN  - [ZKC-connect-executor-0-SendThread(example-client.default.svc.cluster.local:2181):ClientCnxn$SendThread@1112] - Client session timed out, have not heard from server in 11655ms for sessionid 0x1000052f01f050a
2019-03-13 10:33:16,839 - INFO  - [ZKC-connect-executor-0-SendThread(example-client.default.svc.cluster.local:2181):ClientCnxn$SendThread@1160] - Client session timed out, have not heard from server in 11655ms for sessionid 0x1000052f01f050a, closing socket connection and attempting reconnect
2019-03-13 10:33:16,839 - ERROR - [BookieJournal-3181:BookieCriticalThread@41] - Uncaught exception in thread BookieJournal-3181 and is exiting!
java.lang.OutOfMemoryError: Java heap space
2019-03-13 10:33:16,867 - ERROR - [ForceWriteThread:Journal$ForceWriteThread@494] - I/O exception in ForceWrite thread
java.nio.channels.ClosedChannelException
        at sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:110)
        at sun.nio.ch.FileChannelImpl.force(FileChannelImpl.java:379)
        at org.apache.bookkeeper.bookie.BufferedChannel.forceWrite(BufferedChannel.java:201)
        at org.apache.bookkeeper.bookie.JournalChannel.forceWrite(JournalChannel.java:256)
        at org.apache.bookkeeper.bookie.Journal$ForceWriteRequest.process(Journal.java:359)
        at org.apache.bookkeeper.bookie.Journal$ForceWriteThread.run(Journal.java:480)
2019-03-13 10:33:18,030 - INFO  - [ZKC-connect-executor-0-EventThread:ZooKeeperWatcherBase@134] - ZooKeeper client is disconnected from zookeeper now, but it is OK unless we received EXPIRED event.```

### PR DESCRIPTION
**Pull request by adrianmo**
### Change log description

- Store BK index directory in a PVC
- Set Pod `request` and `limit` resources
- Set JVM heap memory values
- Revert `BK_useHostNameAsBookieID` to `false` 

### Purpose of the change
Fix #134 
Fix #137 

### How to test the change

Build an operator image based on this branch and run Pravega system tests.

---
Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

### CODE REVIEW

**Comment by adrianmo** [comment]
[BookKeeper errors on Kubernetes](https://trello.com/c/ihm5UFIc/147-bookkeeper-errors-on-kubernetes)

File: pkg/controller/pravega/bookie.go [hunk 1]
```diff
@@ -167,14 +172,20 @@ func makeBookieVolumeClaimTemplates(spec *v1alpha1.BookkeeperSpec) []corev1.Pers
...
+		},
 	}
 }
 
 func MakeBookieConfigMap(pravegaCluster *v1alpha1.PravegaCluster) *corev1.ConfigMap {
 	configData := map[string]string{
 		"BK_BOOKIE_EXTRA_OPTS":     "\"-Xms1g -Xmx1g -XX:MaxDirectMemorySize=1g -XX:+UseG1GC  -XX:MaxGCPauseMillis=10 -XX:+ParallelRefProcEnabled -XX:+UnlockExperimentalVMOptions -XX:+AggressiveOpts -XX:+DoEscapeAnalysis -XX:ParallelGCThreads=32 -XX:ConcGCThreads=32 -XX:G1NewSizePercent=50 -XX:+DisableExplicitGC -XX:-ResizePLAB\"",
 		"ZK_URL":                   pravegaCluster.Spec.ZookeeperUri,
-		"BK_useHostNameAsBookieID": "true",
+		"BK_useHostNameAsBookieID": "false",
```
Comment (Tristan1900) [inline pkg/controller/pravega/bookie.go]: Can I ask why are we changing this back to `false`?

Comment (adrianmo) [reply pkg/controller/pravega/bookie.go]: We changed this to `true` when https://github.com/pravega/pravega/pull/3357 was merged, however, `pravega/pravega:latest` is still based on BK 4.5, so we will have a regression (https://github.com/pravega/pravega-operator/issues/116) if we keep it `true`. I think we should set it to `false` until Pravega 0.5 is released. 

@Tristan1900 any thoughts?

Comment (Tristan1900) [reply pkg/controller/pravega/bookie.go]: You are right. Thanks for the explanation.

**Pull request merged**